app/recipes/tests/test_recipe_api.py

```diff
@@ -0,0 +1,178 @@
+        self.client = APIClient()
+
+    def test_retrieve_recipes(self):
+        """Test retrieving a list of recipes."""
+        create_recipe()
+        create_recipe()
+
+        res = self.client.get(RECIPES_URL)
+
+        recipes = Recipe.objects.all()
+        serializer = RecipeSerializer(recipes, many=True)
+        self.assertEqual(res.status_code, status.HTTP_200_OK)
+        self.assertEqual(res.data, serializer.data)
+
+    def test_get_recipe_detail(self):
+        """Test get recipe detail."""
+        recipe = create_recipe()
+
+        url = detail_url(recipe.id)
+        res = self.client.get(url)
+
+        serializer = RecipeSerializer(recipe)
+        self.assertEqual(res.data, serializer.data)
+
+    def test_create_recipe(self):
+        """Test creating a recipe."""
+        payload = {
+            'name': 'Sample recipe',
+            'description': 'A good recipe',
+        }
+        res = self.client.post(RECIPES_URL, payload)
+
+        self.assertEqual(res.status_code, status.HTTP_201_CREATED)
+        recipe = Recipe.objects.get(id=res.data['id'])
+        for k, v in payload.items():
+            self.assertEqual(getattr(recipe, k), v)
+
+    def test_partial_update(self):
+        """Test partial update of a recipe."""
+        original_description = 'https://example.com/recipe.pdf'
+        recipe = create_recipe(
+            name='Sample recipe name',
+            description=original_description,
+        )
+
+        payload = {'name': 'New recipe name'}
+        url = detail_url(recipe.id)
+        res = self.client.patch(url, payload)
+
+        self.assertEqual(res.status_code, status.HTTP_200_OK)
+        recipe.refresh_from_db()
+        self.assertEqual(recipe.name, payload['name'])
+        self.assertEqual(recipe.description, original_description)
+
+    def test_full_update(self):
+        """Test full update of recipe."""
+        recipe = create_recipe(
+            name='Sample recipe name',
+            description='Sample recipe description.',
+        )
+
+        payload = {
+            'name': 'New recipe name',
+            'description': 'New recipe description',
+        }
+        url = detail_url(recipe.id)
+        res = self.client.put(url, payload)
+
+        self.assertEqual(res.status_code, status.HTTP_200_OK)
+        recipe.refresh_from_db()
+        for k, v in payload.items():
+            self.assertEqual(getattr(recipe, k), v)
+
+    def test_delete_recipe(self):
+        """Test deleting a recipe successful."""
+        recipe = create_recipe()
+
+        url = detail_url(recipe.id)
+        res = self.client.delete(url)
+
+        self.assertEqual(res.status_code, status.HTTP_204_NO_CONTENT)
+        self.assertFalse(Recipe.objects.filter(id=recipe.id).exists())
+
+    def test_create_recipe_with_existing_ingredient(self):
+        """Test creating a new recipe with existing ingredient."""
+        ingredient = Ingredient.objects.create(name='Lemon')
```

Comment:
to demonstrate what I meant, try creating 2 ingredients here with the name `Lemon` and see how many ingredients you have with this name after the `POST` call
`Ingredient.objects.filter(name='Lemon).count()`